pattern/constraint/thing/RelationConstraint.java

```diff
@@ -177,7 +177,8 @@ public boolean equals(Object o) {
             if (this == o) return true;
             if (o == null || getClass() != o.getClass()) return false;
             final RolePlayer that = (RolePlayer) o;
-            return (Objects.equals(this.roleType, that.roleType)) && (this.player.equals(that.player));
+            return (Objects.equals(this.roleType, that.roleType)) && (this.player.equals(that.player)) &&
+                    this.roleTypeHints.equals(that.roleTypeHints);
```

Comment:
be careful with this - the hash hasn't changed... how is this going to play with hashing overall if we update type hitnts?

Reply:
OK I've moved it back and we'll discuss this later.